rasa/utils/tensorflow/rasa_layers.py

```diff
@@ -0,0 +1,757 @@
+import tensorflow_addons as tfa
+
+from rasa.core.constants import DIALOGUE
+from rasa.shared.nlu.constants import TEXT
+from rasa.utils.tensorflow.model_data import FeatureSignature
+from rasa.utils.tensorflow.constants import (
+    REGULARIZATION_CONSTANT,
+    WEIGHT_SPARSITY,
+    NUM_TRANSFORMER_LAYERS,
+    TRANSFORMER_SIZE,
+    NUM_HEADS,
+    UNIDIRECTIONAL_ENCODER,
+    KEY_RELATIVE_ATTENTION,
+    VALUE_RELATIVE_ATTENTION,
+    MAX_RELATIVE_POSITION,
+    MASKED_LM,
+    HIDDEN_LAYERS_SIZES,
+    DROP_RATE,
+    SPARSE_INPUT_DROPOUT,
+    DENSE_INPUT_DROPOUT,
+    DENSE_DIMENSION,
+    CONCAT_DIMENSION,
+    DROP_RATE_ATTENTION,
+    SEQUENCE,
+    SENTENCE,
+)
+from rasa.utils.tensorflow import layers
+from rasa.utils.tensorflow.transformer import TransformerEncoder
+
+
+# TODO: use this? it's in layers.py
+tfa.options.TF_ADDONS_PY_OPS = True
+
+
+class ConcatenateSparseDenseFeatures(tf.keras.layers.Layer):
+    """Combines multiple sparse and dense feature tensors into one dense tensor.
+
+    This layer combines features from various featurisers into a single feature array
+    per input example. All features must be of the same feature type, i.e. sentence-
+    level or sequence-level (token-level).
+
+    A given list of tensors (whether sparse or dense) is turned into one tensor by:
+    1. converting sparse tensors into dense ones
+    2. optionally, applying dropout to sparse tensors before and/or after the conversion
+    3. concatenating all tensors along the last dimension
+
+    Arguments:
+        attribute: Name of attribute (e.g. `text` or `label`) whose features will be
+            processed.
+        feature_type: Feature type to be processed -- `sequence` or `sentence`.
+        attribute_signature: A list of `FeatureSignature`s for the given attribute and
+            feature type.
+        config: A model config for correctly parametrising the layer.
+
+    Input shape:
+        List of N-D tensors, each with shape: `(batch_size, ..., input_dim)`.
+        All tensors must have the same shape, except the last dimension.
+        All sparse tensors must have the same shape including the last dimension.
+
+    Output shape:
+        N-D tensor with shape: `(batch_size, ..., units)` where `units` is the sum of
+        the last dimension sizes across all input sensors, with sparse tensors instead
+        contributing `config[DENSE_DIMENSION][attribute]` units each.
+
+    Raises:
+        A ValueError if no feature signatures are provided.
+    """
+
+    def __init__(
+        self,
+        attribute: Text,
+        feature_type: Text,
+        attribute_signature: List[FeatureSignature],
+        config: Dict[Text, Any],
+    ) -> None:
+        super().__init__(
+            name=f"concatenate_sparse_dense_features_{attribute}_{feature_type}"
+        )
+
+        if not attribute_signature:
+            raise ValueError("No feature signatures found!")
```

Comment:
That's a very fair point. In general, the way I (don't) check for invalid argument values is very poor. I guess it makes sense to remove this one check, but we'll have to think carefully about what we need to check because otherwise the models could fail in all kinds of weird ways, producing confusing errors...

Reply:
To come back to your question: Given the current code, I think this can't happen.

Reply:
yeah, we cannot add a check for every arg. I think it should be clear that if the arg doesn't have default value, it should be provided